usr/opt/OSC200/usr/share/man/man3/getc.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: getc.3,v $ $Revision: 3.2.5.3 $ (DEC) $Date: 1993/11/29 19:07:28 $
.\"
.\" (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
.\" ALL RIGHTS RESERVED 
.\" OSF/1 1.2
.\"
.TH getc 3
.SH NAME
.PP
\*Lgetc\*O, \*Lfgetc\*O, \*Lgetc_unlocked\*O, \*Lgetchar\*O, \*Lgetchar_unlocked\*O, \*Lgetw\*O \- Get a character or word from an input stream
.SH LIBRARY
.PP
Standard C Library (\*Llibc.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <stdio.h>
.PP
\*Lint getc(
.nL
        \*LFILE \(**\*Vstream\*L);\*O
.PP
\*Lint fgetc(
.nL
        \*LFILE *\*Vstream\*L);\*O
.PP
\*Lint getc_unlocked\*O(
.nL
        \*LFILE * \*Vstream\*L);\*O
.PP
\*Lint getchar(void);\*O
.PP
\*Lint getchar_unlocked(void);\*O
.PP
\*Lint getw(
.nL
        \*LFILE *\*Vstream\*L);\*O
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vstream\*O"
Points to the file structure of an open file.
.LE
.SH DESCRIPTION
.PP
.iX "getc function"
.iX "fgetc function"
.iX "getchar function"
.iX "getw function"
.iX "getc_unlocked function"
.iX "getchar_unlocked function"
.iX "input stream" "getting character from"
.iX "input stream" "getting word from"
The \*Lgetc(\|)\*O function  returns the next byte from the input specified
by the \*Vstream\*O parameter and moves the file pointer, if defined,
ahead one byte in\*V stream\*O. The \*Lgetc(\|)\*O function may be
a macro (depending on compile-time definitions). See the
\*LNOTES\*O section for more information.
.PP
The \*Lfgetc(\|)\*O function performs the same function as
\*Lgetc(\|)\*O.
.PP
The \*Lgetchar(\|)\*O function returns the next byte from \*Lstdin\*O,
the standard input stream. Note that \*Lgetchar(\|)\*O can also be a macro.
.PP
The reentrant versions of these functions are all locked against
multiple threads calling them simultaneously.  This will incur an
overhead to ensure integrity of the stream. The unlocked versions of
these calls, \*Lgetc_unlocked(\|)\*O and \*Lgetchar_unlocked(\|)\*O
may be used to avoid the overhead. The \*Lgetc_unlocked(\|)\*O
and \*Lgetchar_unlocked(\|)\*O functions are functionally identical to the
\*Lgetc(\|)\*O and \*Lgetchar(\|)\*O functions, except that
\*Lgetc_unlocked(\|)\*O and \*Lgetchar_unlocked(\|)\*O
may be safely used only within a scope that
is protected by the \*Lflockfile(\|)\*O and \*Lfunlockfile(\|)\*O functions
used as a pair.
The caller must ensure that the stream is locked before these
functions are used. The \*Lgetc(\|)\*O and \*Lgetchar(\|)\*O
functions can also be macros.
.PP
The \*Lgetw(\|)\*O function returns the next word (\*Lint\*O) from
the input specified by the \*Vstream\*O parameter and increments
the associated file pointer, if defined, to point to the next \*Lint\*O.
The size of an \*Lint\*O varies from one machine architecture to another.
The \*Lgetw(\|)\*O function returns the constant \*LEOF\*O at
the end of the file or when an error occurs. Since \*LEOF\*O is
a valid integer value, the \*Lfeof(\|)\*O and \*Lferror(\|)\*O functions
can be used to check the success of \*Lgetw(\|)\*O. The \*Lgetw(\|)\*O
function assumes no special alignment in the file.
.PP
Because of possible differences in \*Lint\*O length and byte ordering
from one machine architecture to another, files
written using the \*Lputw(\|)\*O
subroutine are machine dependent and may not be readable using \*Lgetw(\|)\*O
on a different type of processor.
.SH NOTES
.VL
.LI "\*LAES Support Level:\*O"
Full use (\*Lgetc(\|)\*O,
\*Lfgetc(\|)\*O,
\*Lgetchar(\|)\*O).
.nL
Trial use (\*Lgetw(\|)\*O).
.LE
.ne 2i
.SH RETURN VALUES
.PP
These functions and macros return the integer constant \*LEOF\*O
at the end of the file or upon an error.
.SH ERRORS
.PP
If any of the following conditions occur, the
\*fgetc(\|)\*O, \*Lgetc(\|)\*O, \*Lgetc_unlocked(\|)\*O, \*Lgetchar(\|)\*O,
\*Lgetchar_unlocked(\|)\*O, and \*Lgetw(\|)\*O functions set
\*Lerrno\*O to the corresponding value:
.VL
.LI "[EAGAIN]"
The \*LO_NONBLOCK\*O flag is set for the underlying \*Vstream\*O and
the process would be delayed by the read operation.
.LI "[EBADF]"
The file descriptor underlying the \*Vstream\*O is not a valid file
descriptor or is not open for reading.
.LI "[EINTR]"
The read operation was interrupted by a signal which was caught and no
data was transferred.
.LI "[EIO]"
The call is attempting to read from the process's controlling terminal
and either the process is ignoring or blocking the \*LSIGTTIN\*O
signal or the process group is orphaned.
.LI "[ENOMEM]"
Insufficient memory is available for the operation.
.LI "[ENXIO]"
Device associated with \*Vstream\*O does not exist.
.LE
.SH RELATED INFORMATION
.PP
Functions:
\*Lflockfile\*O(3),
\*Lfunlockfile\*O(3),
\*Lgets\*O(3),
\*Lgetwc\*O(3),
\*Lputc\*O(3).
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/getc.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: getc.c,v $ $Revision: 4.3.8.2 $ (DEC) $Date: 1993/06/07 22:57:13 $";
#endif 
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
#if !defined(lint) && !defined(_NOIDENT)

#endif
/*
 * FUNCTIONS: getc 
 *
 * (C) COPYRIGHT International Business Machines Corp. 1989 
 * All Rights Reserved
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 * getc.c	1.8  com/lib/c/io,3.1,8943 9/12/89 18:21:51
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#if defined(_THREAD_SAFE)
#pragma weak getc_unlocked = __getc_unlocked
#endif
#endif
#include <errno.h>
#include <stdio.h>
#include "ts_supp.h"

#ifdef	_THREAD_SAFE
#include "stdio_lock.h"
#undef getc_unlocked
#ifdef _NAME_SPACE_WEAK_STRONG
#define getc_unlocked __getc_unlocked
#endif
#endif	/* _THREAD_SAFE */
#undef getc

/*
 * FUNCTION:	A subroutine version of the macro getc.  This function was
 *		created to meet ANSI C standards.  The getc function obtains
 *		the next character (if present) as an unsigned char
 *		converted to an int, from the file pointed to by stream, and
 *		advances the associated file position indicator for the stream.
 *		POSIX 1003.4a defines this to be locked by default so the
 *		single thread version becomes getc_unlocked() and a new
 *		thread safe version is defined with the same interface.
 *
 * RETURN VALUE DESCRIPTION:	
 *		Returns the next character from the input stream pointed to
 *		by stream.  If the stream is at end-of-file, the end-of-file
 *		indicator for the stream is set and getc returns EOF.  If a 
 *		read error occurs, the error indicator for the stream is set 
 *		and getc returns EOF.
 *
 */  

#ifdef _THREAD_SAFE
int	
getc_unlocked(FILE *stream)
#else
int	
getc(FILE *stream)
#endif	/* _THREAD_SAFE */
{
	if ((stream->_flag&(_IOREAD|_IORW))==0) {  /* not open for reading */
		TS_SETERR(EBADF);
		return(EOF);
	}
	if (--(stream)->_cnt < 0)
		return(__filbuf(stream));
	else
		return((int) *(stream)->_ptr++);
}

#ifdef _THREAD_SAFE
int	
getc(FILE *stream)
{
	register int        rc;
	TS_FDECLARELOCK(filelock)

        TS_FLOCK(filelock, stream);
	if ((stream->_flag&(_IOREAD|_IORW))==0) {  /* not open for reading */
		TS_SETERR(EBADF);
                TS_FUNLOCK(filelock);
		return(EOF);
	}
	if (--(stream)->_cnt < 0)
		rc = __filbuf(stream);
	else
		rc = (int) *(stream)->_ptr++;
	TS_FUNLOCK(filelock);
	return(rc);
}
#endif	/* _THREAD_SAFE */
